probe-rs/src/architecture/arm/sequences/atsame5x.rs

```diff
@@ -0,0 +1,379 @@
+    pub dccd0, _: 2;
+    // Debugger Present
+    // This bit is set when a debugger probe is detected.
+    // This bit is never cleared.
+    pub dbgpres, _: 1;
+    // Protected
+    // This bit is set at power-up when the device is protected.
+    // This bit is never cleared.
+    pub prot, _: 0;
+
+}
+
+impl From<u8> for DsuStatusB {
+    fn from(value: u8) -> Self {
+        Self(value)
+    }
+}
+
+impl From<DsuStatusB> for u8 {
+    fn from(value: DsuStatusB) -> Self {
+        value.0
+    }
+}
+
+impl DsuStatusB {
+    const ADDRESS: u64 = 0x4100_2102;
+}
+
+/// A wrapper for different types that can perform SWD Commands (SWJ_Pins SWJ_Sequence)
+struct SwdSequenceShim<'a>(&'a mut dyn architecture::arm::communication_interface::DapProbe);
+
+impl<'a> From<&'a mut dyn architecture::arm::communication_interface::DapProbe>
+    for SwdSequenceShim<'a>
+{
+    fn from(p: &'a mut dyn architecture::arm::communication_interface::DapProbe) -> Self {
+        Self(p)
+    }
+}
+
+impl<'a> architecture::arm::communication_interface::SwdSequence for SwdSequenceShim<'a> {
+    fn swj_sequence(&mut self, bit_len: u8, bits: u64) -> Result<(), Error> {
+        self.0.swj_sequence(bit_len, bits).map_err(Error::Probe)
+    }
+
+    fn swj_pins(&mut self, pin_out: u32, pin_select: u32, pin_wait: u32) -> Result<u32, Error> {
+        self.0
+            .swj_pins(pin_out, pin_select, pin_wait)
+            .map_err(Error::Probe)
+    }
+}
+
+/// Marker struct indicating initialization sequencing for ATSAM D5x/E5x family parts.
+pub struct AtSAME5x {}
+
+impl AtSAME5x {
+    /// Create the sequencer for the ATSAM D5x/E5x family of parts.
+    pub fn create() -> Arc<Self> {
+        Arc::new(Self {})
+    }
+
+    /// Perform a Chip-Erase operation
+    ///
+    /// # Errors
+    /// This operation can fail due to insufficient permissions, or if Chip-Erase Lock is
+    /// enabled (this lock can only be released from within the device firmare).
+    /// After a successful Chip-Erase a `DebugProbeError::ReAttachRequired` is returned
+    /// to signal that a re-connect is needed for the DSU to start operating in unlocked mode.
+    pub fn erase_all(&self, memory: &mut Memory, permissions: &Permissions) -> Result<(), Error> {
+        let dsu_status_b = DsuStatusB::from(memory.read_word_8(DsuStatusB::ADDRESS)?);
+
+        match (dsu_status_b.celck(), dsu_status_b.prot(), permissions.erase_all()) {
+            (true, _, _) => Err(Error::MissingPermissions(
+                "Chip-Erase is locked. This can only be unlocked from within the device firmware by performing \
+                a Chip-Erase Unlock (CEULCK) command."
+                    .into(),
+            )),
+            (false, true, Err(_)) => Err(Error::MissingPermissions(
+                "Device is locked. A Chip-Erase operation is required to unlock. \
+                            Re-run with '--connect-under-reset --allow-erase-all' to perform Chip-Erase."
```

Comment:
This is not a good choice of words because the flags do not exist in the lib and are not present in all binary tools. I am not sure how to formulate this better tho.
Maybe `Re-run with granting the 'erase-all' permission and connecting under reset.`

Reply:
Good point, I'll re-phrase